coderd/database/dbfake/dbfake.go

```diff
@@ -0,0 +1,83 @@
+package dbfake
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/database/dbgen"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/coder/v2/coderd/provisionerdserver"
+	"github.com/coder/coder/v2/coderd/telemetry"
+	sdkproto "github.com/coder/coder/v2/provisionersdk/proto"
+)
+
+// CreateWorkspace inserts a workspace into the database.
+func CreateWorkspace(t testing.TB, db database.Store, seed database.Workspace) database.Workspace {
+	t.Helper()
+
+	// This intentionally fulfills the minimum requirements of the schema.
+	// Tests can provide a custom template ID if necessary.
+	if seed.TemplateID == uuid.Nil {
+		template := dbgen.Template(t, db, database.Template{
+			OrganizationID: seed.OrganizationID,
+			CreatedBy:      seed.OwnerID,
+		})
+		seed.TemplateID = template.ID
+	}
+	return dbgen.Workspace(t, db, seed)
+}
```

Comment:
The seed OwnerID and OrgID might not match the workspace. If the `OwnerID`is omitted, it will be a random uuid and the workspace will not match.

The easy thing is to do what you did for the templateID.

```golang
seed.TemplateID = template.ID
seed.OwnerID = template.CreatedBy
seed.OrganizationID = template.OrganizationID
```

The other option that I do not know if it will work because it requires the constraints to be deferable, is to do it in a TX.

```golang
	// This might not work....
	db.InTx(func(store database.Store) error {
		seed = dbgen.Workspace(t, db, seed)
		template := dbgen.Template(t, db, database.Template{
			ID: seed.TemplateID,
			OrganizationID: seed.OrganizationID,
			CreatedBy:      seed.OwnerID,
		})
			db.InTx(func(store database.Store) error {
		seed = dbgen.Workspace(t, db, seed)
		template := dbgen.Template(t, db, database.Template{
			ID: seed.TemplateID,
			OrganizationID: seed.OrganizationID,
			CreatedBy:      seed.OwnerID,
		})
		
		return nil
	}, nil)
		return nil
	}, nil)
```

Reply:
I mention this because the arguments are pretty vague as to which fields are required to be populated. So we could also solve this by adding some failure conditions.

```golang
require.NotEqualf(t, uuid.Nil, seed.OrganizationID, "seed.OrganizationID must be set")
// ... other required fields.
```

Reply:
Fixed in the latest push I believe!